File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/OrderService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.OrderRefundDto;
import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.SetOrderToBoughtDto;
import at.ac.tuwien.sepm.groupphase.backend.entity.Order;

import java.util.List;

public interface OrderService {

    /**
     * Method for returning all reserved orders from the user.
     *
     * @param userEmail of the order.
     * @return all the orders stored in database
     */
    List<Order> getAllReserved(String userEmail);

    /**
     * Method for returning all bought orders from the user.
     *
     * @param userEmail of the order.
     * @return all the orders stored in database
     */
    List<Order> getAllBought(String userEmail);

    /**
     * Method for returning all orders from the user.
     *
     * @param userEmail of the order.
     * @return all the orders stored in database
     */
    List<Order> getAll(String userEmail);

    /**
     * Method for setting an order to bought and update the paymentInformation for the order.
     *
     * @param setOrderToBoughtDto with the id´s of order and paymentInformation
     */
    void setOrderToBought(SetOrderToBoughtDto setOrderToBoughtDto);

    /**
     * Refunds a bought or reserved order.
     *
     * @param orderRefundDto containg orderId
     */
    void refund(OrderRefundDto orderRefundDto);
}

File: frontend/src/app/services/order.service.ts
import {Injectable} from '@angular/core';
import {Globals} from '../global/globals';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Order} from '../dtos/order';
import {SetOrderToBoughtDto} from '../dtos/setOrderToBoughtDto';

@Injectable({
  providedIn: 'root'
})
export class OrderService {
  private baseUri: string = this.globals.backendUri + '/orders';

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  /**
   * Gets all reserved orders from logged in user.
   *
   * @param email of the user.
   */
  getReservedOrders(email: string): Observable<Order[]> {
    console.log('Load all reserved orders');
    return this.httpClient.get<Order[]>(this.baseUri + '/' + email + '/reserved');
  }

  /**
   * Gets all bought orders from logged in user.
   *
   * @param email of the user.
   */
  getBoughtOrders(email: string): Observable<Order[]> {
    console.log('Load all bought orders');
    return this.httpClient.get<Order[]>(this.baseUri + '/' + email + '/bought');
  }

  /**
   * Gets all orders from logged in user.
   *
   * @param email of the user.
   */
  getAllOrders(email: string): Observable<Order[]> {
    console.log('Load all orders');
    return this.httpClient.get<Order[]>(this.baseUri + '/' + email);
  }

  /**
   * Method for setting an order to bought and update the paymentInformation for the order.
   *
   * @param setOrderToBoughtDto
   */
  setOrderToBought(setOrderToBoughtDto: SetOrderToBoughtDto): Observable<string> {
    console.log('Set order to bought');
    return this.httpClient.put(this.baseUri, setOrderToBoughtDto, {responseType: 'text'});
  }

  refundOrder(orderId: number): Observable<void> {
    return this.httpClient.put<void>(this.baseUri + '/refund', {orderId});
  }
}
